### ISSUE TITLE
Host unhandled exception when function does not handle a CancellationToken

### ISSUE THREAD

**Issue by liliankasem** [Azure/azure-functions-host#8978]
Edit: See latest comment for up-to-date information about this issue

Issue found in a [CRI](https://portal.microsofticm.com/imp/v3/incidents/details/354815015/home).

When we send a cancellation request, and the function we are invoking does not handle the cancellation token, the host should be handling any exceptions/errors thrown. During the CRI investigation I found that in this scenario we get an unhandled host error.

```kusto
// EUS region
FunctionsLogs
| where TIMESTAMP >= datetime(2022-12-12T15:42:38.6393661Z) and TIMESTAMP <= datetime(2022-12-12T15:43:38.6393661Z)
| where AppName =~ 'APP_NAME' or AppName startswith 'APP_NAME'
| where RoleInstance has "pd1SmallDedicatedWebWorkerRole_IN_12980"
| project-reorder TIMESTAMP, HostVersion, RoleInstance, FunctionName, Summary, FunctionInvocationId
| order by TIMESTAMP asc
```

> See CRI link for app name and more details

RoleInstance: pd1SmallDedicatedWebWorkerRole_IN_12980

![Screenshot 2022-12-14 at 2 57 11 PM](https://user-images.githubusercontent.com/2198905/207733403-23d27048-2055-4c7b-98d9-e5bdfb4831b1.png)

The function invocation failure exception: 

```log
System.Threading.Tasks.TaskCanceledException : A task was canceled.
   at async Microsoft.Azure.WebJobs.Script.Description.WorkerFunctionInvoker.InvokeCore(Object[] parameters,FunctionInvocationContext context) at /_/src/WebJobs.Script/Description/Workers/WorkerFunctionInvoker.cs : 96
```

The error details we see for "unhandled host error":

```log
System.Threading.Tasks.TaskCanceledException : A task was canceled.
   at async Microsoft.Azure.WebJobs.Script.Description.WorkerFunctionInvoker.InvokeCore(Object[] parameters,FunctionInvocationContext context) at /_/src/WebJobs.Script/Description/Workers/WorkerFunctionInvoker.cs : 96
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.Azure.WebJobs.Script.Description.FunctionInvokerBase.Invoke(Object[] parameters) at /_/src/WebJobs.Script/Description/FunctionInvokerBase.cs : 82
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.Azure.WebJobs.Script.Description.FunctionGenerator.Coerce[T](Task`1 src) at /_/src/WebJobs.Script/Description/FunctionGenerator.cs : 225
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.Azure.WebJobs.Host.Executors.FunctionInvoker`2.InvokeAsync[TReflected,TReturnValue](Object instance,Object[] arguments) at C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionInvoker.cs : 52
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.InvokeWithTimeoutAsync(IFunctionInvoker invoker,ParameterHelper parameterHelper,CancellationTokenSource timeoutTokenSource,CancellationTokenSource functionCancellationTokenSource,Boolean throwOnTimeout,TimeSpan timerInterval,IFunctionInstance instance) at C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionExecutor.cs : 581
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.ExecuteWithWatchersAsync(IFunctionInstanceEx instance,ParameterHelper parameterHelper,ILogger logger,CancellationTokenSource functionCancellationTokenSource) at C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionExecutor.cs : 527
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.ExecuteWithLoggingAsync(IFunctionInstanceEx instance,FunctionStartedMessage message,FunctionInstanceLogEntry instanceLogEntry,ParameterHelper parameterHelper,ILogger logger,CancellationToken cancellationToken) at C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionExecutor.cs : 306
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.ExecuteWithLoggingAsync(IFunctionInstanceEx instance,FunctionStartedMessage message,FunctionInstanceLogEntry instanceLogEntry,ParameterHelper parameterHelper,ILogger logger,CancellationToken cancellationToken) at C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionExecutor.cs : 352
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.Azure.WebJobs.Host.Executors.FunctionExecutor.TryExecuteAsync(IFunctionInstance functionInstance,CancellationToken cancellationToken) at C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\FunctionExecutor.cs : 108
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.Azure.WebJobs.Host.Executors.ExceptionDispatchInfoDelayedException.Throw() at C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\Executors\ExceptionDispatchInfoDelayedException.cs : 28
   at async Microsoft.Azure.WebJobs.JobHost.CallAsyncCore(IFunctionDefinition function,Object functionKey,IDictionary`2 arguments,CancellationToken cancellationToken) at C:\projects\azure-webjobs-sdk-rqm4t\src\Microsoft.Azure.WebJobs.Host\JobHost.cs : 264
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.Azure.WebJobs.Script.WebHost.Features.FunctionExecutionFeature.ExecuteAsync(HttpRequest request,CancellationToken cancellationToken) at /_/src/WebJobs.Script.WebHost/Features/FunctionExecutionFeature.cs : 74
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.Azure.WebJobs.Script.WebHost.Middleware.FunctionInvocationMiddleware.GetResultAsync(HttpContext context,IFunctionExecutionFeature functionExecution) at /_/src/WebJobs.Script.WebHost/Middleware/FunctionInvocationMiddleware.cs : 132
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.Azure.WebJobs.Script.WebHost.Middleware.FunctionInvocationMiddleware.Invoke(HttpContext context) at /_/src/WebJobs.Script.WebHost/Middleware/FunctionInvocationMiddleware.cs : 54
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at async Microsoft.Azure.WebJobs.Script.WebHost.Middleware.ExceptionMiddleware.Invoke(HttpContext context) at /_/src/WebJobs.Script.WebHost/Middleware/ExceptionMiddleware.cs : 27,
```

> Note: 8 different role instances received the cancellation request around the same time (due to an app restart), but not all of them got a host unhandled exception

- [x] Investigate why we are getting an unhandled host error and handle the exception to fix this issue

**Comment by liliankasem** [comment]
This might actually be an issue with the dotnet worker. When a function invocation we are trying to cancel throws `TaskCanceledException` exception, we are supposed to be catching that and returning a valid invocation response with the state set to cancelled:

[Code](https://github.com/Azure/azure-functions-dotnet-worker/blob/09b8bde8693edf5109498fbfdf2135efa6efd17d/src/DotNetWorker.Grpc/Handlers/InvocationHandler.cs#L111-L120)

```csharp
            catch (Exception ex)
            {
                response.Result.Exception = enableUserCodeException ? ex.ToUserRpcException() : ex.ToRpcException();
                response.Result.Status = StatusResult.Types.Status.Failure;

                if (ex.InnerException is TaskCanceledException or OperationCanceledException)
                {
                    response.Result.Status = StatusResult.Types.Status.Cancelled;
                }
            }
```

**Comment by liliankasem** [comment]
Upon further investigation we found that this may not actually be a bug. The order of events is correct.

The invocation with ID `9841ae49-d99b-42fc-bd35-dae6c8fd4e01` that we wanted to cancel (`Sending invocation cancel request for InvocationId 9841ae49-d99b-42fc-bd35-dae6c8fd4e01`) did get a cancellation request but because it finished in time, the function executed successfully and did not need to be cancelled.

And for the invocation with ID `77ee7054-b447-4967-ab45-f61f60fdbd7f`, it did not get a cancellation request in the first place because we handled that early on by not sending the invocation in the first place: `Cancellation has been requested, cancelling invocation request` This log message is supposed to indicate that we are cancelling the invocation request for `77ee7054-b447-4967-ab45-f61f60fdbd7f`.

So far this all lines up. The behaviour is as expected but we can improve logging here to make things more clear:

- [x] Update logs for `Cancellation has been requested, cancelling invocation request` to include the invocation ID and make clear that the InvocationRequest is not being sent, and not that we are cancelling the invocation
- [x] Move the cancellation token registration to after the `await SendStreamingMessageAsync ` code block to avoid sending a CancellationRequest before an InvocationRequest
    - [code](https://github.com/Azure/azure-functions-host/blob/fe291a2030a902dadb4db98951a2013851e068ac/src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs#L647-L650)
